src/renderer/grid_renderer.rs

```diff
@@ -209,4 +193,79 @@ impl GridRenderer {
 
         canvas.restore();
     }
+
+    fn draw_underline(
+        &self,
+        canvas: &mut Canvas,
+        style: &Arc<Style>,
+        underline_style: UnderlineStyle,
+        p1: Point,
+        p2: Point,
+    ) {
+        canvas.save();
+
+        let mut underline_paint = self.paint.clone();
+        let auto_scaling = SETTINGS.get::<RendererSettings>().underline_automatic_scaling;
+        // Arbitrary value under which we simply round the line thickness to 1. Anything else
+        // results in ugly aliasing artifacts.
+        let stroke_width = if self.shaper.current_size() < 15. || auto_scaling == false {
+            underline_paint.set_anti_alias(false);
+            1.0
+        } else {
+            underline_paint.set_anti_alias(true);
+            self.shaper.current_size() / 10.
+        };
+
+        underline_paint
+            .set_color(style.special(&self.default_style.colors).to_color())
+            .set_stroke_width(stroke_width);
+
+        match underline_style {
+            UnderlineStyle::Underline => {
+                underline_paint.set_path_effect(None);
+                canvas.draw_line(p1, p2, &underline_paint);
+            }
+            UnderlineStyle::UnderDouble => {
+                underline_paint.set_path_effect(None);
+                canvas.draw_line(p1, p2, &underline_paint);
+                let p1 = (p1.x, p1.y - 2 as f32);
+                let p2 = (p2.x, p2.y - 2 as f32);
+                canvas.draw_line(p1, p2, &underline_paint);
+            }
+            UnderlineStyle::UnderCurl => {
+                let p1 = (p1.x, p1.y - 2.);
+                let p2 = (p2.x, p2.y - 2.);
+                underline_paint
+                    .set_path_effect(None)
+                    .set_anti_alias(true)
+                    .set_style(skia_safe::paint::Style::Stroke);
+                let mut path = Path::default();
+                path.move_to(p1);
+                let mut i = p1.0;
+                let mut sin = -2. * stroke_width;
+                while i <= p2.0 {
+                    sin *= -1.;
+                    i += 4. * stroke_width;
+                    path.quad_to((i - 2. * stroke_width, p1.1 + sin), (i, p1.1));
+                }
```

Comment:
To be honest, this could maybe get more precise. Instead of doing the last "half" curve, we could maybe try to clip it or stop beforehand and use `p2.0` as the final X coordinate of the last target point, plus some arcsin hackery on the Y coordinate. But that might or might not look ugly.

Reply:
Absolutely, I had that planned and then completely forgot to do it. Clipping is easy enough, but it may look awkward on lines where multiple consecutive fragments have different undercurls.

An easy fix would be to simply have the width of the wave be the same as the font width. That way it'll repeat seamlessly and we don't need to clip it anymore.

Reply:
There's the result:
![olpD34x](https://user-images.githubusercontent.com/4097716/184456200-ba87a219-90b7-4942-ac8d-216ae25874ab.png)